File: acs-integration-tests/src/test/java/com/ge/predix/integration/test/PolicyEvaluationWithAttributeConnectorIT.java
/*******************************************************************************
 * Copyright 2017 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *******************************************************************************/

package com.ge.predix.integration.test;

import java.io.IOException;
import java.net.URI;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;

import org.apache.commons.lang.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.cloud.sleuth.DefaultSpanNamer;
import org.springframework.cloud.sleuth.Span;
import org.springframework.cloud.sleuth.TraceKeys;
import org.springframework.cloud.sleuth.instrument.web.HttpTraceKeysInjector;
import org.springframework.cloud.sleuth.instrument.web.ZipkinHttpSpanInjector;
import org.springframework.cloud.sleuth.instrument.web.client.TraceRestTemplateInterceptor;
import org.springframework.cloud.sleuth.log.NoOpSpanLogger;
import org.springframework.cloud.sleuth.sampler.AlwaysSampler;
import org.springframework.cloud.sleuth.trace.DefaultTracer;
import org.springframework.cloud.sleuth.util.ArrayListSpanAccumulator;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.ImportResource;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.oauth2.client.OAuth2RestTemplate;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
import org.springframework.web.client.RestTemplate;
import org.testng.Assert;
import org.testng.annotations.AfterClass;
import org.testng.annotations.BeforeClass;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

import com.ge.predix.acs.model.Effect;
import com.ge.predix.acs.rest.AttributeAdapterConnection;
import com.ge.predix.acs.rest.AttributeConnector;
import com.ge.predix.acs.rest.PolicyEvaluationRequestV1;
import com.ge.predix.acs.rest.PolicyEvaluationResult;
import com.ge.predix.acs.rest.Zone;
import com.ge.predix.test.utils.ACSITSetUpFactory;
import com.ge.predix.test.utils.ACSTestUtil;
import com.ge.predix.test.utils.PolicyHelper;
import com.ge.predix.test.utils.ZoneFactory;

@Configuration
@ImportResource("classpath:integration-test-spring-context.xml")
class PolicyEvaluationWithAttributeConnectorITConfiguration {

    @Autowired
    private ACSITSetUpFactory acsitSetUpFactory;

    @Autowired
    private DefaultTracer tracer;

    private void setRestTemplateInterceptor(final RestTemplate restTemplate) {
        restTemplate.setInterceptors(Collections.singletonList(
                new TraceRestTemplateInterceptor(this.tracer, new ZipkinHttpSpanInjector(),
                        new HttpTraceKeysInjector(this.tracer, new TraceKeys()))));
    }

    @Bean
    public DefaultTracer tracer() {
        return new DefaultTracer(new AlwaysSampler(), new Random(), new DefaultSpanNamer(), new NoOpSpanLogger(),
                new ArrayListSpanAccumulator(), new TraceKeys());
    }

    @Bean
    public OAuth2RestTemplate acsAdminRestTemplate() throws IOException {
        OAuth2RestTemplate acsAdminRestTemplate;
        this.acsitSetUpFactory.setUp();
        acsAdminRestTemplate = this.acsitSetUpFactory.getAcsAdminRestTemplate2();
        setRestTemplateInterceptor(acsAdminRestTemplate);
        return acsAdminRestTemplate;
    }
    
    @Bean
    public ACSITSetUpFactory acsitSetUpFactory() {
        return this.acsitSetUpFactory;
    }
}

@ContextConfiguration(classes = { PolicyEvaluationWithAttributeConnectorITConfiguration.class })
public class PolicyEvaluationWithAttributeConnectorIT extends AbstractTestNGSpringContextTests {

    @Value("${ZONE1_NAME:testzone1}")
    private String acsZone1Name;

    @Value("${ACS_UAA_URL}")
    private String acsUaaUrl;

    @Value("${ADAPTER_ENDPOINT:${ASSET_ADAPTER_URL}}")
    private String adapterEndpoint;

    @Value("${ADAPTER_UAA_TOKEN_URL:${ASSET_TOKEN_URL}}")
    private String adapterUaaTokenUrl;

    @Value("${ADAPTER_UAA_CLIENT_ID:${ASSET_CLIENT_ID}}")
    private String adapterUaaClientId;

    @Value("${ADAPTER_UAA_CLIENT_SECRET:${ASSET_CLIENT_SECRET}}")
    private String adapterUaaClientSecret;

    @Autowired
    private ZoneFactory zoneFactory;

    @Autowired
    private PolicyHelper policyHelper;

    @Autowired
    private OAuth2RestTemplate acsAdminRestTemplate;

    @Autowired
    private ACSITSetUpFactory acsitSetUpFactory;

    @Autowired
    private DefaultTracer tracer;

    private static final String TEST_PART_ID = "part/03f95db1-4255-4265-a509-f7bca3e1fee4";

    private Zone zone;

    private URI resourceAttributeConnectorUrl;

    private HttpHeaders zoneHeader() throws IOException {
        HttpHeaders httpHeaders = ACSTestUtil.httpHeaders();
        httpHeaders.set(PolicyHelper.PREDIX_ZONE_ID, this.zone.getSubdomain());
        return httpHeaders;
    }

    private void configureAttributeConnector(final boolean isActive) throws IOException {

        List<AttributeAdapterConnection> adapters = Collections.singletonList(
                new AttributeAdapterConnection(this.adapterEndpoint, this.adapterUaaTokenUrl, this.adapterUaaClientId,
                        this.adapterUaaClientSecret));

        AttributeConnector attributeConnector = new AttributeConnector();
        attributeConnector.setIsActive(isActive);
        attributeConnector.setAdapters(new HashSet<>(adapters));
        this.acsAdminRestTemplate.exchange(this.resourceAttributeConnectorUrl, HttpMethod.PUT,
                new HttpEntity<>(attributeConnector, zoneHeader()), AttributeConnector.class);
    }

    @BeforeClass
    void beforeClass() throws IOException {

        this.zone = this.acsitSetUpFactory.getZone1();
        this.resourceAttributeConnectorUrl = URI.create(this.zoneFactory.getAcsBaseURL() + "/v1/connector/resource");
    }

    private void deconfigureAttributeConnector() throws IOException {
        this.acsAdminRestTemplate
                .exchange(this.resourceAttributeConnectorUrl, HttpMethod.DELETE, new HttpEntity<>(zoneHeader()),
                        Void.class);
    }

    @AfterClass
    void afterClass() throws IOException {
        this.acsitSetUpFactory.destroy();
    }

    @Test(dataProvider = "adapterStatusesAndResultingEffects")
    public void testPolicyEvaluationWithAdapters(final boolean adapterActive, final Effect expectedEffect,
            final boolean enableSleuthTracing) throws Exception {
        String testPolicyName = this.policyHelper
                .setTestPolicy(this.acsAdminRestTemplate, zoneHeader(), this.zoneFactory.getAcsBaseURL(),
                        "src/test/resources/policy-set-with-one-policy-using-resource-attributes-from-asset-adapter"
                                + ".json");

        try {
            this.configureAttributeConnector(adapterActive);
            PolicyEvaluationRequestV1 policyEvaluationRequest = this.policyHelper
                    .createEvalRequest("GET", "testSubject", TEST_PART_ID, null);

            if (enableSleuthTracing) {
                this.tracer.continueSpan(Span.builder().traceId(1L).spanId(2L).parent(3L).build());
            }

            ResponseEntity<PolicyEvaluationResult> policyEvaluationResponse = this.acsAdminRestTemplate
                    .postForEntity(this.zoneFactory.getAcsBaseURL() + PolicyHelper.ACS_POLICY_EVAL_API_PATH,
                            new HttpEntity<>(policyEvaluationRequest, zoneHeader()), PolicyEvaluationResult.class);
            Assert.assertEquals(policyEvaluationResponse.getStatusCode(), HttpStatus.OK);

            HttpHeaders responseHeaders = policyEvaluationResponse.getHeaders();
            Assert.assertTrue(
                    responseHeaders.containsKey(Span.TRACE_ID_NAME) && StringUtils.isNotEmpty(Span.TRACE_ID_NAME));
            if (enableSleuthTracing) {
                Assert.assertEquals(Span.hexToId(responseHeaders.get(Span.TRACE_ID_NAME).get(0)), 1L);
            }

            PolicyEvaluationResult policyEvaluationResult = policyEvaluationResponse.getBody();
            Assert.assertEquals(policyEvaluationResult.getEffect(), expectedEffect);
        } finally {
            this.policyHelper
                    .deletePolicySet(this.acsAdminRestTemplate, this.zoneFactory.getAcsBaseURL(), testPolicyName,
                            zoneHeader());
            this.deconfigureAttributeConnector();
        }
    }

    @DataProvider
    private Object[][] adapterStatusesAndResultingEffects() {
        return new Object[][] { { true, Effect.PERMIT, true }, { true, Effect.PERMIT, false },
                { false, Effect.NOT_APPLICABLE, true } };
    }
}

File: acs-integration-tests/src/test/java/org/eclipse/keti/integration/test/PolicyEvaluationWithAttributeConnectorIT.kt
/*******************************************************************************
 * Copyright 2018 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.keti.integration.test

import org.apache.commons.lang.StringUtils
import org.eclipse.keti.acs.model.Effect
import org.eclipse.keti.acs.rest.AttributeAdapterConnection
import org.eclipse.keti.acs.rest.AttributeConnector
import org.eclipse.keti.acs.rest.PolicyEvaluationResult
import org.eclipse.keti.acs.rest.Zone
import org.eclipse.keti.test.utils.ACSITSetUpFactory
import org.eclipse.keti.test.utils.ACSTestUtil
import org.eclipse.keti.test.utils.PolicyHelper
import org.eclipse.keti.test.utils.ZoneFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.cloud.sleuth.DefaultSpanNamer
import org.springframework.cloud.sleuth.Span
import org.springframework.cloud.sleuth.TraceKeys
import org.springframework.cloud.sleuth.instrument.web.HttpTraceKeysInjector
import org.springframework.cloud.sleuth.instrument.web.ZipkinHttpSpanInjector
import org.springframework.cloud.sleuth.instrument.web.client.TraceRestTemplateInterceptor
import org.springframework.cloud.sleuth.log.NoOpSpanLogger
import org.springframework.cloud.sleuth.sampler.AlwaysSampler
import org.springframework.cloud.sleuth.trace.DefaultTracer
import org.springframework.cloud.sleuth.util.ArrayListSpanAccumulator
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.ImportResource
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.security.oauth2.client.OAuth2RestTemplate
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests
import org.springframework.web.client.RestTemplate
import org.testng.Assert
import org.testng.annotations.AfterClass
import org.testng.annotations.BeforeClass
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import java.io.IOException
import java.net.URI
import java.util.HashSet
import java.util.Random

@Configuration
@ImportResource("classpath:integration-test-spring-context.xml")
internal class PolicyEvaluationWithAttributeConnectorITConfiguration {

    @Autowired
    private lateinit var acsitSetUpFactory: ACSITSetUpFactory

    @Autowired
    private lateinit var tracer: DefaultTracer

    private fun setRestTemplateInterceptor(restTemplate: RestTemplate) {
        restTemplate.interceptors = listOf<ClientHttpRequestInterceptor>(
            TraceRestTemplateInterceptor(
                this.tracer, ZipkinHttpSpanInjector(),
                HttpTraceKeysInjector(this.tracer, TraceKeys())
            )
        )
    }

    @Bean
    fun tracer(): DefaultTracer {
        return DefaultTracer(
            AlwaysSampler(), Random(), DefaultSpanNamer(), NoOpSpanLogger(),
            ArrayListSpanAccumulator(), TraceKeys()
        )
    }

    @Bean
    @Throws(IOException::class)
    fun acsAdminRestTemplate(): OAuth2RestTemplate {
        val acsAdminRestTemplate: OAuth2RestTemplate = this.acsitSetUpFactory.acsZonesAdminRestTemplate
        this.acsitSetUpFactory.setUp()
        setRestTemplateInterceptor(acsAdminRestTemplate)
        return acsAdminRestTemplate
    }

    @Bean
    fun acsitSetUpFactory(): ACSITSetUpFactory? {
        return this.acsitSetUpFactory
    }
}

private const val TEST_PART_ID = "part/03f95db1-4255-4265-a509-f7bca3e1fee4"

@ContextConfiguration(classes = [(PolicyEvaluationWithAttributeConnectorITConfiguration::class)])
class PolicyEvaluationWithAttributeConnectorIT : AbstractTestNGSpringContextTests() {

    @Value("\${ACS_UAA_URL}")
    private lateinit var acsUaaUrl: String

    @Value("\${ADAPTER_ENDPOINT:\${ASSET_ADAPTER_URL}}")
    private lateinit var adapterEndpoint: String

    @Value("\${ADAPTER_UAA_TOKEN_URL:\${ASSET_TOKEN_URL}}")
    private lateinit var adapterUaaTokenUrl: String

    @Value("\${ADAPTER_UAA_CLIENT_ID:\${ASSET_CLIENT_ID}}")
    private lateinit var adapterUaaClientId: String

    @Value("\${ADAPTER_UAA_CLIENT_SECRET:\${ASSET_CLIENT_SECRET}}")
    private lateinit var adapterUaaClientSecret: String

    @Autowired
    private val zoneFactory: ZoneFactory? = null

    @Autowired
    private val policyHelper: PolicyHelper? = null

    @Autowired
    private val acsAdminRestTemplate: OAuth2RestTemplate? = null

    @Autowired
    private val acsitSetUpFactory: ACSITSetUpFactory? = null

    @Autowired
    private val tracer: DefaultTracer? = null

    private var zone: Zone? = null

    private var resourceAttributeConnectorUrl: URI? = null

    @Throws(IOException::class)
    private fun zoneHeader(): HttpHeaders {
        val httpHeaders = ACSTestUtil.httpHeaders()
        httpHeaders.set(PolicyHelper.PREDIX_ZONE_ID, this.zone!!.subdomain)
        return httpHeaders
    }

    @Throws(IOException::class)
    private fun configureAttributeConnector(isActive: Boolean) {

        val adapters = listOf(
            AttributeAdapterConnection(
                this.adapterEndpoint, this.adapterUaaTokenUrl, this.adapterUaaClientId,
                this.adapterUaaClientSecret
            )
        )

        val attributeConnector = AttributeConnector()
        attributeConnector.isActive = isActive
        attributeConnector.adapters = HashSet(adapters)
        this.acsAdminRestTemplate!!.exchange(
            this.resourceAttributeConnectorUrl, HttpMethod.PUT,
            HttpEntity(attributeConnector, zoneHeader()), AttributeConnector::class.java
        )
    }

    @BeforeClass
    @Throws(IOException::class)
    internal fun beforeClass() {

        this.zone = this.acsitSetUpFactory!!.zone1
        this.resourceAttributeConnectorUrl = URI.create(this.zoneFactory!!.acsBaseURL + "/v1/connector/resource")
    }

    @Throws(IOException::class)
    private fun deconfigureAttributeConnector() {
        this.acsAdminRestTemplate!!
            .exchange(
                this.resourceAttributeConnectorUrl, HttpMethod.DELETE, HttpEntity<Any>(zoneHeader()),
                Void::class.java
            )
    }

    @AfterClass
    @Throws(IOException::class)
    internal fun afterClass() {
        this.acsitSetUpFactory!!.destroy()
    }

    @Test(dataProvider = "adapterStatusesAndResultingEffects")
    @Throws(Exception::class)
    fun testPolicyEvaluationWithAdapters(
        adapterActive: Boolean,
        expectedEffect: Effect,
        enableSleuthTracing: Boolean
    ) {
        val testPolicyName = this.policyHelper!!
            .setTestPolicy(
                this.acsAdminRestTemplate, zoneHeader(), this.zoneFactory!!.acsBaseURL,
                "src/test/resources/policy-set-with-one-policy-using-resource-attributes-from-asset-adapter" + ".json"
            )

        try {
            this.configureAttributeConnector(adapterActive)
            val policyEvaluationRequest = this.policyHelper
                .createEvalRequest("GET", "testSubject", TEST_PART_ID, null)

            if (enableSleuthTracing) {
                this.tracer!!.continueSpan(Span.builder().traceId(1L).spanId(2L).parent(3L).build())
            }

            val policyEvaluationResponse = this.acsAdminRestTemplate!!
                .postForEntity(
                    this.zoneFactory.acsBaseURL + PolicyHelper.ACS_POLICY_EVAL_API_PATH,
                    HttpEntity(policyEvaluationRequest, zoneHeader()), PolicyEvaluationResult::class.java
                )
            Assert.assertEquals(policyEvaluationResponse.statusCode, HttpStatus.OK)

            val responseHeaders = policyEvaluationResponse.headers
            Assert.assertTrue(
                responseHeaders.containsKey(Span.TRACE_ID_NAME) && StringUtils.isNotEmpty(Span.TRACE_ID_NAME)
            )
            if (enableSleuthTracing) {
                Assert.assertEquals(Span.hexToId(responseHeaders[Span.TRACE_ID_NAME]?.get(0)), 1L)
            }

            val policyEvaluationResult = policyEvaluationResponse.body
            Assert.assertEquals(policyEvaluationResult.effect, expectedEffect)
        } finally {
            this.policyHelper
                .deletePolicySet(
                    this.acsAdminRestTemplate, this.zoneFactory.acsBaseURL, testPolicyName,
                    zoneHeader()
                )
            this.deconfigureAttributeConnector()
        }
    }

    @DataProvider
    private fun adapterStatusesAndResultingEffects(): Array<Array<Any?>> {
        return arrayOf<Array<Any?>>(
            arrayOf(true, Effect.PERMIT, true),
            arrayOf(true, Effect.PERMIT, false),
            arrayOf(false, Effect.NOT_APPLICABLE, true)
        )
    }
}
